src/runtime/range.c

```diff
@@ -90,6 +90,39 @@ boolean rangemap_insert_range(rangemap rm, range r)
     return true;
 }
 
+/* If the hole is in the middle of the range of a node, splits the node in two.
+ * Returns false if the hole is not contained in any existing node or if a new node cannot be
+ * allocated.
+ */
+boolean rangemap_insert_hole(rangemap rm, range r)
+{
+    struct rmnode k = {
+        .r = r,
+    };
+    rangemap_foreach_of_range(rm, curr, &k) {
+        if ((curr->r.start > r.start) || (curr->r.end < r.end))
+            return false;
+        if (curr->r.start < r.start) {
+            if (curr->r.end > r.end) {
+                rmnode n = allocate(rm->h, sizeof(*n));
+                if (n == INVALID_ADDRESS)
+                    return false;
+                rmnode_init(n, irange(r.end, curr->r.end));
+                curr->r.end = r.start;
+                rangemap_insert(rm, n);
+            } else {
+                curr->r.end = r.start;
+            }
+        } else if (curr->r.end > r.end) {
+            curr->r.start = r.end;
```

Comment:
Changing curr->r should a rangemap_reinsert(). In this particular case, it's possible that the adjustment of start would not actually cause any reordering in the underlying data structure (rbtree), but this isn't a guarantee at the interface level.

Similarly, changing curr->r.end above should not effect a change in the tree - but, again, there could be subtle breakage should the rangemap implementation be changed in the future.